redhat-appstudio/remote-secret
# Uploading webhook

### What does this PR do?

Adds ability to upload secretData along with RemoteSecret creating/updating using Mutating Webhook.

Note: This PR is not going to be merged as is anyway since it requires change in SecretStorage API (using UID as a part of SecretData identifier blocks using it by design, see https://issues.redhat.com/projects/SVPI/issues/SVPI-550 )

### What issues does this PR fix or reference?

It significantly improves ability of dynamic SecretData uploading. Comparing to using Upload Secret (which is NOT excluded anyway) it:
- increases security of secret data (it never appeared in etcd)
- improves UX and supportability, using sync REST call in a webhook allows reacting immediately (user creating the Secret will get clear error message instead of analyzing error CR status   

### How to test this PR?
Build/push/deploy as usual.
Put SecretData to the extra-field RemoteSecret->data as an array of key/value pairs, like:

```
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: myrs1
  namespace: default
data:
  - key: username
    value: Z2VuYQ==
  - key: password
    value: Z2VuYQ==

```

Create RemoteSecret and check:
- RS called myrs1 is created on default namespace and does not contain 'data' field
- remotesecret remotesecret-controller-manager-XXXX log contains something like:

> {"level":"info","ts":"2023-07-18T06:42:32.160Z","caller":"webhook/remotesecret_mutator.go:34","msg":"Data DETECTED, upload it and delete from here","Data":{"password":"Z2VuYQ==","username":"Z2VuYQ=="}}        
> {"level":"info","ts":"2023-07-18T06:42:32.160Z","caller":"webhook/remotesecret_mutator.go:56","msg":"Data STORED ","data":{"password":"Z2VuYQ==","username":"Z2VuYQ=="}}                                          
Which  means data is stored and read from the SecretStorage.
